### PR TITLE
proper testable Ruby Event logging and Java Event logging support

### PR DESCRIPTION
**Pull request by colinsurprenant**
this PR builds on PR #4261 which cleans the Event class and adds the following:

- replace the Ruby `Event` `LOGGER` constant with an injectable logger object which allows correct testing as discussed in issue #4253 
- adds testable logging support in the Java `Event` and makes the Java `Event` pass all Ruby `Event` specs.

### CODE REVIEW

File: logstash-core-event-java/spec/event_spec.rb [hunk 1]
```diff
@@ -135,4 +135,54 @@
       expect(event["tags"]).to eq(["foo"])
     end
   end
+
+  module DummyLogger
```
Comment (purbon) [inline logstash-core-event-java/spec/event_spec.rb]: I would move this to a support location and rename it somehow including the word stub/mock. Moving this to a support location will help reusing this loggers in case of others need it.

File: logstash-core-event-java/spec/event_spec.rb [hunk 2]
```diff
@@ -135,4 +135,54 @@
...
+    end
+  end
+
+  context "logger" do
+
+    let(:logger) { double("Logger") }
+    after(:each) {  LogStash::Event.logger = nil }
+
+    it "should set logger using a module" do
+      LogStash::Event.logger = DummyLogger
```
Comment (purbon) [inline logstash-core-event-java/spec/event_spec.rb]: I don't understand what is the purpose of this test if the next one uses a rspec mock?

File: logstash-core-event-java/spec/event_spec.rb [hunk 3]
```diff
@@ -135,4 +135,54 @@
...
+
+      # then unset
+      LogStash::Event.logger = nil
+      expect(logger).to receive(:warn).never
+      # this will produce a log line in stdout by the Java Event
+      LogStash::Event.new(TIMESTAMP => "ignore this log")
+    end
+
+
+    it "should warn on parsing errort" do
```
Comment (purbon) [inline logstash-core-event-java/spec/event_spec.rb]: typo?

**Comment by purbon** [comment]
I only see test for the invalid situations, can we add some test for valid outcomes? so we test both sides of the story.

File: logstash-core-event/spec/logstash/event_spec.rb [hunk 4]
```diff
@@ -354,8 +359,13 @@
       end
 
       it "should warn for invalid string format" do
-        expect(LogStash::Event::LOGGER).to receive(:warn)
+        logger_mock = double("logger")
```
Comment (purbon) [inline logstash-core-event/spec/logstash/event_spec.rb]: I see some duplication in here? I know is not in the style of the spec, but we might introduce this now and make usage of let statements as you do for the core-event-java spec.

**Comment by purbon** [comment]
I put some comments on specs, but in general the idea looks good!

File: logstash-core-event-java/spec/event_spec.rb (hunk 1 above)
Comment (colinsurprenant) [inline logstash-core-event-java/spec/event_spec.rb]: this is really specific to this test and the stage the implementation is in now, i.e.: only the warn method implemented. it should stay here.

File: logstash-core-event-java/spec/event_spec.rb (hunk 2 above)
Comment (colinsurprenant) [inline logstash-core-event-java/spec/event_spec.rb]: good point, I will add a bit of documentation, but the idea is that passing a real class/module vs a mock uses different objects and the JRuby ext implementation has to be compatible with both.

**Comment by colinsurprenant** [comment]
@purbon for invalid/valid test: the specs in here are typically only for specific cases related to the Java & JRuby ext implementation that either were required while doing the implementation or are really specific for corner cases for only this implementation. Note that all logstash-core-event specs must and are passing this implementation

File: logstash-core-event-java/spec/event_spec.rb [hunk 5]
```diff
@@ -135,4 +135,62 @@
...
+
+    it "should set logger using a module" do
+      LogStash::Event.logger = DummyLogger
+      expect(DummyLogger).to receive(:warn).once
+      LogStash::Event.new(TIMESTAMP => "invalid timestamp")
+    end
+
+    it "should set logger using a mock" do
+      LogStash::Event.logger = logger
+      expect(logger).to receive(:warn)
```
Comment (jsvd) [inline logstash-core-event-java/spec/event_spec.rb]: since in line 160 you use .once, maybe do it here also?

**Comment by jsvd** [comment]
Minor comment, otherwise LGTM.
Is there an issue to remove the technical debt of having two spec suites
https://github.com/colinsurprenant/logstash/blob/fix/event_logging/logstash-core-event/spec/logstash/event_spec.rb

**Comment by colinsurprenant** [comment]
@jsvd it is a good point. some specs might be similar but the ones in the Java implementation were, for most part, added to validate this specific implementation. At some point it might be a good idea to review this and possibly merge/refactor them but this will be done at a later stage and we'll have to think about general specs vs contract specs, etc. I don't think having "more" specs is necessarily too harmful of tech debt so for me this is not too urgent for now.

File: logstash-core-event/spec/logstash/event_spec.rb [hunk 6]
```diff
@@ -340,10 +340,15 @@
       end
 
       it "should warn for invalid value" do
-        expect(LogStash::Event::LOGGER).to receive(:warn).twice
+        LogStash::Event.logger = logger_mock
```
Comment (jordansissel) [inline logstash-core-event/spec/logstash/event_spec.rb]: Feels like this should be done in a `before` and the resetting `...logger = nil` should be in an `after` ?

Comment (colinsurprenant) [inline logstash-core-event/spec/logstash/event_spec.rb]: +1, it feels like that :D

File: logstash-core-event-java/src/main/java/com/logstash/Event.java [hunk 7]
```diff
@@ -28,9 +28,9 @@
     public static final String VERSION = "@version";
     public static final String VERSION_ONE = "1";
 
-    private static final ObjectMapper mapper = new ObjectMapper();
-
-    // TODO: add metadata support
+    private static final Logger DEFAULT_LOGGER = new StdioLogger();
+    private transient final ObjectMapper mapper = new ObjectMapper();
+    private transient static Logger logger = DEFAULT_LOGGER;
```
Comment (talevy) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: if this is being set at runtime, why should this be `static`?

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java [hunk 8]
```diff
@@ -274,5 +286,12 @@ public IRubyObject ruby_tag(ThreadContext context, RubyString value)
         public IRubyObject ruby_timestamp(ThreadContext context) throws IOException {
             return new JrubyTimestampExtLibrary.RubyTimestamp(context.getRuntime(), this.event.getTimestamp());
         }
+
+        @JRubyMethod(name = "logger=", required = 1, meta = true)
+        public static IRubyObject ruby_set_logger(ThreadContext context, IRubyObject recv, IRubyObject value)
+        {
+            Event.setLogger(value.isNil() ? null : new ProxyLogger((RubyObject)value));
```
Comment (talevy) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: should we fail this if it is `null`? I understand that the current behavior internally is to default to the default logger (stdio), but technically this operation failed since it could not set a new logger.

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 7 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: because its a *class* variable, the same logger for all Event instances.

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java (hunk 8 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: good question. not sure, will think about it.

File: logstash-core-event-java/spec/event_spec.rb [hunk 9]
```diff
@@ -135,4 +135,62 @@
...
+
+    let(:logger) { double("Logger") }
+    after(:each) {  LogStash::Event.logger = nil }
+
+    # the following 2 specs are using both a real module (DummyLogger)
+    # and a mock. both tests are needed to make sure the implementation
+    # supports both types of objects.
+
+    it "should set logger using a module" do
+      LogStash::Event.logger = DummyLogger
```
Comment (jordansissel) [inline logstash-core-event-java/spec/event_spec.rb]: Why not use the double here?

Comment (jordansissel) [inline logstash-core-event-java/spec/event_spec.rb]: I suppose my question is more that - why test both module and mock?

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 7 above)
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: Should we use a singleton or factory pattern here instead of a global variable? Feels weird to have a mutable global variable.

File: logstash-core-event-java/src/main/java/com/logstash/Event.java [hunk 10]
```diff
@@ -225,18 +225,16 @@ private Timestamp initTimestamp(Object o) {
             } else if (o instanceof RubySymbol) {
                 return new Timestamp(((RubySymbol) o).asJavaString());
             } else {
-                // TODO: add logging
-                //return Timestamp.now();
-                throw new IllegalArgumentException();
+                Event.logger.warn("Unrecognized " + TIMESTAMP + " value type=" + o.getClass().toString());
```
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: Can we use structured logging instead of string formatting?

File: logstash-core-event-java/src/main/java/com/logstash/Event.java [hunk 11]
```diff
@@ -250,4 +248,8 @@ public void tag(String tag) {
             tags.add(tag);
         }
     }
+
+    public static void setLogger(Logger logger) {
+        Event.logger = (logger == null) ? DEFAULT_LOGGER : logger;
```
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: Should we allow null? If DEFAULT_LOGGER is a constant, what if instead there's a setLogger(DEFAULT_LOGGER) or resetLoggerToDefault() ?

File: logstash-core-event-java/src/main/java/com/logstash/Logger.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+// minimalist Logger interface to wire a logger callback in the Event class
+// for now only warn is defined because this is the only method that's required
+// in the Event class.
+// TODO: (colin) generalize this
+
+public interface Logger {
+
+    // TODO: (colin) complete interface beyond warn when needed
+
+    void warn(String message);
```
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Logger.java]: Can we do structured logging similar to Cabin or log4j's MDC?

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java [hunk 13]
```diff
@@ -1,15 +1,10 @@
 package com.logstash.ext;
 
-import com.logstash.Event;
-import com.logstash.PathCache;
-import com.logstash.RubyToJavaConverter;
-import com.logstash.Timestamp;
+import com.logstash.*;
```
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: Why use wildcard imports? Explicit imports, for me at least, make it easier to read.

File: logstash-core-event-java/spec/event_spec.rb (hunk 9 above)
Comment (colinsurprenant) [inline logstash-core-event-java/spec/event_spec.rb]: as written in the comment above the test, this is needed to avoid regression basically. That is because of the fast that this is going through the JRuby Java Ext wrapper which at first did not work with the mock/double. so I had to adjust the code to be more permissive on the type of object it would accept.

maybe the comment is not clear or detailed enough for wht "why"...?

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java (hunk 8 above)
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: I am also in favor of avoiding nil/null where possible.

File: logstash-core-event/lib/logstash/event.rb [hunk 14]
```diff
@@ -245,16 +243,24 @@ def fields
     raise DeprecatedMethod
   end
 
+  def self.logger=(logger)
+    @@logger = logger
```
Comment (jordansissel) [inline logstash-core-event/lib/logstash/event.rb]: Are class variables in JRuby thread-safe?

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 11 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: that's in line with @talevy comment above for the same thing but in the JRubu wrapper. I need to think about it.

File: logstash-core-event-java/src/main/java/com/logstash/Logger.java (hunk 12 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Logger.java]: we can, this is the minimal implementation to support what is currently used in the Event. I did not want to go through a complete logging infrastructure work at this point, but basically this is a passthrough/proxy to whatever logging strategy is used in Ruby.

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java (hunk 13 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: agree. this is IntelliJ reorganizing the imports...

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 7 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: I am not sure. at this point this is only mutated in the tests/specs. there is not really *currently* any point in changing the logger at runtime. for now this is what make it work with the less overhead, this is certainly something we can improve.

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 10 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: yes, I guess we could. I would prefer to create a new PR for this though.

File: logstash-core-event-java/src/main/java/com/logstash/Logger.java (hunk 12 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Logger.java]: I see your point now related to the other comment above (I was going backward in the comment :P). So yes, I agree, but would prefer to push back that improvement in another PR.

**Comment by colinsurprenant** [comment]
after this first pass of reviews I realize I probably should have done 2 separate PRs, one for the current Ruby stuff and the other for the Java Event because both have different expectations for the reviews:
- the Ruby code need to be reviewed the "normal" way for merging into master

- the Java Event code is still not production quality and does not have the same quality expectations. this is code to get me going to an initial/base version that will pass **all** specs. this code **is** merged in master but is not in any *live* execution path and will not be included in the release package until we have the ready stamp on it. I suggest that any *bigger improvements* be noted and created as issues so I/we can work on them at a later stage.

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java (hunk 13 above)
Comment (talevy) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: it is possible to configure intellij to not do this. I think you have to set a threshold class count before wildcarding (and just set it to something crazy).

File: logstash-core-event/lib/logstash/event.rb (hunk 14 above)
Comment (colinsurprenant) [inline logstash-core-event/lib/logstash/event.rb]: I am not sure that thread safety is a *real* problem here:
- `@@logger` is only assigned once normally. We currently do not reassign loggers at runtime.
- that being said, `@@logger` could probably be initialized sooner instead of in the the first `#Event#logger` call
- in JRuby defining or modifying constants and class variables is thread safe

Are there any other concerns I missed?

File: logstash-core-event-java/spec/event_spec.rb (hunk 9 above)
Comment (jordansissel) [inline logstash-core-event-java/spec/event_spec.rb]: your comment is clear, I didn't read it somehow. haha - thanks! <3

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 7 above)
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: I'm OK with a global variable here if we comment that it's intended for testing/specs.

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 10 above)
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: +1 on waiting for a new PR for log structuring.

File: logstash-core-event-java/src/main/java/com/logstash/Logger.java (hunk 12 above)
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/Logger.java]: +1 on your proposal to add structured logging in another PR, not this one.

File: logstash-core-event/lib/logstash/event.rb (hunk 14 above)
Comment (jordansissel) [inline logstash-core-event/lib/logstash/event.rb]: OK let's leave this as-is. I am satisfied ;P

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java (hunk 13 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: yeah, need to do this :P thanks for the heads up

File: logstash-core-event/lib/logstash/event.rb (hunk 14 above)
Comment (colinsurprenant) [inline logstash-core-event/lib/logstash/event.rb]: will add a comment per this discussion though

**Comment by colinsurprenant** [comment]
@ph just a ping since I know you are working on refactoring logging stuff.

File: logstash-core-event-java/spec/event_spec.rb (hunk 5 above)
Comment (colinsurprenant) [inline logstash-core-event-java/spec/event_spec.rb]: +1

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 7 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: +1 adding comments

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 10 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: added issue #4249

File: logstash-core-event-java/src/main/java/com/logstash/Event.java (hunk 11 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/Event.java]: will expose DEFAULT_LOGGER and remove nil/null support

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java (hunk 8 above)
Comment (colinsurprenant) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: will expose DEFAULT_LOGGER and remove nil/null support

File: logstash-core-event/spec/logstash/event_spec.rb (hunk 6 above)
Comment (colinsurprenant) [inline logstash-core-event/spec/logstash/event_spec.rb]: fixed

**Comment by colinsurprenant** [comment]
@jsvd @talevy @jordansissel fixes following your comments, would like another round of review! Thanks!

**Comment by guyboertje** [comment]
It's a pity that I was not pushy enough on this RSpec matcher for Cabin.  https://github.com/jordansissel/ruby-cabin/pull/36

File: logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java [hunk 15]
```diff
@@ -1,47 +1,72 @@
...
                 return new RubyEvent(runtime, rubyClass);
             }
         }, module);
-        clazz.setConstant("LOGGER", runtime.getModule("Cabin").getClass("Channel")
-                .callMethod("get", runtime.getModule("LogStash")));
+
         clazz.setConstant("TIMESTAMP", runtime.newString(Event.TIMESTAMP));
         clazz.setConstant("TIMESTAMP_FAILURE_TAG", runtime.newString(Event.TIMESTAMP_FAILURE_TAG));
         clazz.setConstant("TIMESTAMP_FAILURE_FIELD", runtime.newString(Event.TIMESTAMP_FAILURE_FIELD));
+        clazz.setConstant("DEFAULT_LOGGER", runtime.getModule("Cabin").getClass("Channel").callMethod("get", runtime.getModule("LogStash")));
```
Comment (jordansissel) [inline logstash-core-event-java/src/main/java/com/logstash/ext/JrubyEventExtLibrary.java]: +1 this, good idea.

File: logstash-core-event/spec/logstash/event_spec.rb [hunk 16]
```diff
@@ -314,7 +314,11 @@
     end
 
     context "timestamp initialization" do
-      let(:logger) { double("logger") }
+      let(:logger_mock) { double("logger") }
+
+      after(:each) do
+        LogStash::Event.logger = LogStash::Event::DEFAULT_LOGGER
```
Comment (jordansissel) [inline logstash-core-event/spec/logstash/event_spec.rb]: ++

**Comment by jordansissel** [comment]
Second pass review - LGTM

**Comment by colinsurprenant** [comment]
thanks, will merge!

**Comment by colinsurprenant** [comment]
merged into master and 2.x
